### ISSUE TITLE
Dask pivot_table requires much more memory than pandas pivot_table

### ISSUE THREAD

**Issue by xanderdunn** [dask/dask#6108]
- dask 2.13.0
- Python 3.7.5

I am loading 135 .csv.gz files as a dask dataframe. The CSV files represent 1,341,951,117 rows of data with 5 columns. 3 of the columns are floats, one is a string id, and one of them is a date like '2018-01-01'. Then I set the date column to a categorical. Then pivot the table with one of the columns as the values. Example code:
```python
data = dd.read_csv("/my/path/files_*.csv.gz", compression="gzip", blocksize=None)
data_categorized = data.categorize(columns=["date"]) # This completes in about a minute, uses all CPUs, and does not memory overflow
data_pivoted = data_categorized.pivot_table(index="id", columns="date", values="one_of_the_three_float_columns"
```

I have also tried setting a RAM limit on the distributed client.  I run this on a Google Cloud Compute instance with 956GB of RAM and 40 CPUs and I set the client's limitations accordingly:
```python
from distributed import Client
client = Client(memory_limit='900GB', n_workers=40)
```
Invariably, the dask `pivot_table` consumes all 956GB of RAM and then throws many out of memory errors. On the other hand, when I run this as just a pandas `pivot_table`, it successfully completes without ever using more than 300GB of RAM. It takes a long time, but it does complete. Is dask using far more memory because it's trying to operate on many different partitions of the dataset in parallel? Can I limit the number of parallel operations? I am looking at [some](https://docs.dask.org/en/latest/best-practices.html) [docs](https://stackoverflow.com/questions/44657631/strategy-for-partitioning-dask-dataframes-efficiently) on partitioning, but I'm unsure if it's relevant here.

I can't share the data I use to reproduce this, but I imagine generating a dataset as I described at the beginning would replicate this memory problem. If necessary, I can generate such a dataset, test that it fails in my use case, and post it here.

My use case is very similar to the one described in rapidsai/cudf#1214

**Comment by jsignell** [comment]
It would definitely be helpful to provide a smaller reproducible example that mimics the behavior that you describe. You might also want to  take a look at `data.chunks` and see if you could benefit from any of the approaches described in the chunks docs: https://docs.dask.org/en/latest/array-chunks.html

**Comment by xanderdunn** [comment]
It looks like the chunks API is for Dask arrays?  I'm using Dask Dataframes.  Is there an equivalent to the `pivot_table` functionality for arrays?

**Comment by TomAugspurger** [comment]
>  Is there an equivalent to the pivot_table functionality for arrays?

Nope.

Dask DataFrames are partitioned as well.

Are you converting the date to categorical for a reason? It's a shot in the Dark, but perhaps Dask is expanding the full categorical for every partition, so we many rows for partitions that have very few actual observations? e.g. if partitions `0` has just the dates `2018-01-01`, we'd still end up allocating a DataFrame with all the unique dates, rather than just a single row.

Since you're `aggfunc` is just `mean`, I think you'd be fine doing the pivot_table on the non-categorical data and then reindexing if you really need all the dates in the output.

**Comment by TomAugspurger** [comment]
(A minimal example would be extremely helpful right now: http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports)

**Comment by xanderdunn** [comment]
> if partitions 0 has just the dates 2018-01-01, we'd still end up allocating a DataFrame with all the unique dates, rather than just a single row.

I don't believe this is the case.  The dates in the dataframe are pretty randomly distributed and probably most dates appear in every partition. This could be solved fairly readily though.  I could sort on date first, but that's not something we're currently doing.

> Are you converting the date to categorical for a reason?

The Dataframe `pivot_table` function requires that the column used in the `columns` parameter is a categorical.  If it isn't it throws an error.  See [here](https://github.com/dask/dask/blob/5a7f4524217d0d3ee9e44b6b6513264de7f16171/dask/dataframe/reshape.py#L216). 

> Since you're aggfunc is just mean, I think you'd be fine doing the pivot_table on the non-categorical data and then reindexing if you really need all the dates in the output.

I don't believe the aggfunc comes into play with our data.  It is guaranteed that every (date, id) pair occurs only once. 

@TomAugspurger Thanks for your help 👍

**Comment by xanderdunn** [comment]
In my attempt to create a self-contained code sample I came up with this:
```python
#!/usr/bin/env python3

  import numpy as np
  import pandas as pd
  from dask import dataframe as dd
  from distributed import Client

  if __name__ == '__main__':
      # Create the dataset
      num_unique_ids = int(1e6)
      num_dates = 1300
      dates = np.arange("2016-01-01",
                        np.datetime64("2016-01-01", "D") + np.timedelta64(num_dates, "D"),
                        dtype="datetime64[D]")
      dates_column = np.tile(dates, num_unique_ids)
      print("{:,} dates: ".format(len(dates_column)), dates_column)
      ids = np.arange(0, num_unique_ids)
      ids_column = np.repeat(ids, len(dates))
      print("{:,} ids: ".format(len(ids_column)), ids_column)
      data_column = np.random.choice(int(1e6), len(ids_column))
      print("{:,} data values: ".format(len(data_column)), data_column)
      df = pd.DataFrame({"date": dates_column, "id": ids_column, "count": data_column})
      print(df)
      data = dd.from_pandas(df, npartitions=134) # 134 is the number of different .csv.gz files we load from disk for our real   data
      print(data)

      # Pivot the dataset
      client = Client(memory_limit='900GB', n_workers=40)
      data_categorized = data.categorize(columns=["date"])
      data_pivoted = data_categorized.pivot_table(index="id", columns="date", values="count")
      print(data_pivoted)
```
But it actually successfully pivoted this dataset in less than 2min without ever using more than 160GB of RAM, which is very different from what I see in our real use case.

There are several important differences between this example and our actual use case:
- The `id`s in our real data are alpha-numeric strings of length 12, not `int`s.
- Because of the way I generated the `id`s in the example dataset, they're ordered.  Both the dates and the `id`s are randomly ordered in our dataset, though sorting them is possible.
- The real data is split between 137 .csv.gz's on disk.  We load them directly into dask rather than a pandas dataframe first.

Loading the real dataset into dask from the .csv.gz files reveals that the dtypes are also different:
```python
>>> data = dd.read_csv("files*.csv.gz", compression="gzip", blocksize=None)
>>> data
Dask DataFrame Structure:
                       date      id count_total count_1       count_2
npartitions=135
                     object  object       int64      int64      int64
                        ...     ...         ...        ...        ...
...                     ...     ...         ...        ...        ...
                        ...     ...         ...        ...        ...
                        ...     ...         ...        ...        ...
Dask Name: from-delayed, 405 tasks
```
The `date` column has not been recognized as a date, and as mentioned above the `id` column is a string.

**Comment by TomAugspurger** [comment]
@xanderdunn thanks.

@jsignell brings up a good point: since you're loading from a gzipped file, Dask won't do any partitioning. Can you verify that `data.npartitions` is 1?

I'd recommend somehow splitting your data into multiple partitions: perhaps with a `.repartition` before `categorize`, or perhaps splitting them and then writing them back to disk as a directory of smaller files. Then reading them back in as a glob.

**Comment by xanderdunn** [comment]
This doesn't appear to be the case. 
```python
from dask import dataframe as dd
data = dd.read_csv("my/path/*.csv.gz", compression="gzip", blocksize=None)
```

Result:
```python
>>> data
Dask DataFrame Structure:
                      date           id count_total count_1 count_2
npartitions=135
                     object  object       int64      int64      int64
                        ...     ...         ...        ...        ...
...                     ...     ...         ...        ...        ...
                        ...     ...         ...        ...        ...
                        ...     ...         ...        ...        ...
Dask Name: from-delayed, 405 tasks
```
```python
>>> print(data.npartitions)
135
```

If we imagine it were operating on a single partition, surely its performance should be worst case the same as pandas?  Pandas is operating on "1 partition" and it completes this `pivot_table` using less than 1/3rd the RAM.

**Comment by TomAugspurger** [comment]
Oh sorry, I missed that you had a directory of CSV files rather than a single one. Ignore that then.

---

Yes, your expectation for worst-case sounds correct, but I'm not sure about the details of the implementation.
m
Do we have a minimal example that reproduces the problem? If so I can take a closer look. It sounds like https://github.com/dask/dask/issues/6108#issuecomment-616834257 doesn't reproduce it?

**Comment by xanderdunn** [comment]
The ultimate issue is that a normal pandas `pivot_table` is using far less RAM and is successfully completing, while with the dask read from .csv.gz I can't predict how much RAM it will use and it's running out of RAM and not completing.  The example code I posted generates data with the same number of rows as the real data.  But, unlike my real data, the `pivot_table` of the example data does complete on my machine.  It uses about 150GB of RAM.  Would it succeed on a machine with 100GB of RAM? Would a pandas `pivot_table` of my example data succeed on a machine with 100GB of RAM? The pandas `pivot_table` appears to be much more RAM efficient. This could be expected given that dask's goal is to break it up and work on multiple pieces in parallel.  The real issue then is knowing how many partitions / how much RAM will a dask pivot_table consume?

I believe datatypes are one reason that my example data is succeeding while my real data is not:
- `id` is `object` in my real data and `int64` in my example data
- `date` is `object` in my real data and `datetime64[ns]` in my example data

This fails with my real data and 956GB RAM:
```python
from distributed import Client
client = Client(memory_limit='900GB', n_workers=40)
from dask import dataframe as dd
data = dd.read_csv("/my/path/files_*.csv.gz", compression="gzip", blocksize=None)
data_categorized = data.categorize(columns=["date"])
data_pivoted = data_categorized.pivot_table(index="id", columns="date", values="count") #Out of memory
```

This succeeds with my real data and 956GB RAM:
```python
from distributed import Client
client = Client(memory_limit='900GB', n_workers=40)
from dask import dataframe as dd
# !! Notice the difference: the date column is not dtype datetime64[ns] instead of object
data = dd.read_csv("/my/path/files_*.csv.gz", compression="gzip", blocksize=None, parse_dates=["date"]) 
data_categorized = data.categorize(columns=["date"])
data_pivoted = data_categorized.pivot_table(index="id", columns="date", values="count" # Succeeds
```
However, computing it still runs out of memory:
```python
data_pivoted.compute() # Out of memory and fails
```

I will endeavor to modify my example code so that it exports the generated data to .csv.gz and then I expect the data types will be the same and the issue will reproduce on my machine.  But the minimal amount of data that reproduces the issue will depend on the amount of RAM available.

**Comment by TomAugspurger** [comment]
I don't think we need or want an example that uses that much RAM. Ideally we'll have a small example that can run on a reasonably-sized laptop and monitor the memory usage, and see if anything sticks out as strange.

**Comment by xanderdunn** [comment]
In that case, my example will work for you.  Modify `num_unique_ids` `num_dates` until the dataset is sufficiently small for you.  Compare the memory usage of pandas and dask.

**Comment by pranav-kohli** [comment]
I have a similar setup, Dask dataframe of 25 gb with 32 npartitions and dask workers 4gb each scalable horizontally, meaning I can add workers on the fly. 
I am doing pivot_table on the df and it is giving me worker killed due to memory error. Tried increasing the number of partitions to 394(basically 64 mb chunks) but still the same
Is pivot_table not distributed, does it bring everything to the memory like compute()?

**Comment by jsignell** [comment]
It'd be interesting to know if you are seeing the same behavior when using `groupby` for instance:

```python
import dask

df = dask.datasets.timeseries()
data_categorized = df.categorize(columns="name")
data_categorized = data_categorized.reset_index()
data_categorized["doy"] = data_categorized.timestamp.dt.dayofyear

%timeit data_categorized.pivot_table(index="doy", columns="name", values="x").compute()
%timeit data_categorized.groupby(["doy", "name"])[["x"]].mean().compute().unstack()
```

This is what that looks like for me:

![image](https://user-images.githubusercontent.com/4806877/89227785-4eb81700-d5ac-11ea-9fcd-5e8a77319e16.png)

**Comment by pranav-kohli** [comment]
I will try the above, but the compute would bring back result to the memory, which is what I wanted to avoid.

**Comment by arita37** [comment]
Dask is lacking of benchmark on large pivot tables,
and tends to be buggy (ie necessary of doing category,
need to create own composite index).

**Comment by jsignell** [comment]
For pivot tables dask will try to get the column names before you call compute. That could be resulting in the issues you are seeing.

**Comment by jsignell** [comment]
@pranav-kohli are you also using timestamps as the columns? Also how many categories do you have in the columns field (`len(data_categorized[column].cat.categories)`)?

**Comment by pranav-kohli** [comment]
@jsignell Yes the dataset has a timestamp column, and the total categories is 3

**Comment by jsignell** [comment]
Hmm ok. I really think we probably won't get anywhere on this issue without an [MRE](http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

**Comment by LutzFassl** [comment]
I have the same problem and impression like @pranav-kohli 
(my task is 192Mio rows from CSV file distributed over 104 partitions)
Everything is running smoothly and distributed until I run the pivot_table function. After that I only run to_parquet to execute to task.
The categorical column consists of numbers 1..35 in string format, categorized with 
`ddf["col"] = ddf["col"].astype(str)`
`ddf["col"] = ddf["col"].astype('category')`
`ddf["col"] = ddf["col"].cat.as_known()`

I have 2 workers with each 1 thread at 4GB (local cluster). (It's not a lot of memory allocated, but same result at 2x10GB ram.)
As soon as the dashboard shows the pivoting task the ram blows up to > 20GB (several hundred percent of allocated memory per worker) . It's like I'm running pandas and not dask. Why can it not run distributed? 

I will try to provide a MRE

**Comment by LutzFassl** [comment]
Ok I did the MRE. Pardon me for creating an actual CSV file on your disk, but it seemed to behave differently when I read a numpy array to a ddf directly with from_array. If dask does finish, it will take 30-60 minutes on top of the 10minutes to create the CSV file

<details><summary>CODE, also there is a Jupyter Notebook attached below</summary>
<p>

```python
# First thing to do:
# Set 'terminate' to False to avoid restarting when memory overflows: can better reproduce the problem
# has to be set in distributed.yaml of respective python env

##  creating a dataset
import numpy as np
n_rows_circa = 100000000

# create categorical column made of integers
columns_after_pivot = np.arange(1,36, dtype=np.uint32)
n_cols = len(columns_after_pivot)
n_data_points = n_rows_circa // len(columns_after_pivot)
cols_long = np.tile(columns_after_pivot, n_data_points)
n_rows = len(cols_long)

# create data column
values = np.random.randint(0,4500,n_rows,dtype=np.uint32)

# create index column
start_index = 1856474 # just any first indexnumber
indices_unique = np.arange(start_index,start_index + n_data_points, dtype=np.uint32)
indices = np.repeat(indices_unique, n_cols)

# Build the table from the 3 cols
# Saving to csv takes ~10 minutes
my_big_table = np.c_[indices,cols_long,values]
my_big_table = my_big_table.astype(np.uint32)
np.savetxt("big_table.csv", my_big_table, fmt='%u') #finally creates the 1.5GB testfile

# Unlinking Numpy objects for GC
cols_long, values, indices, my_big_table    = None, None, None, None



# Now start to reproduce the problem



import dask.dataframe as dd
from dask.distributed import Client, LocalCluster
import dask

cluster = LocalCluster(n_workers=2, 
                threads_per_worker=1,
                memory_limit='4000MB'
                )

client = Client(cluster)  # set up local cluster

# self check if terminate==False
print(dask.config.get("distributed.worker.memory"))


ddf = dd.read_csv("big_table.csv", blocksize="14MB" ,header=None, names=["ID","Category","Values"])
# not using from_array because behaved a bit differently ddf = dd.from_array(my_big_table, chunksize=n_rows_circa//104, columns=["ID","Category","Values"])


# change parameter ID to string to help categorize
ddf["Category"] = ddf["Category"].astype(str)
ddf["Category"] = ddf["Category"].astype('category')
ddf["Category"] = ddf["Category"].cat.as_known()


# remove ambiguous indices (which are not present in this test dataset)
ddf = ddf.drop_duplicates(["ID","Category"])

# pivot & save
ddf = ddf.pivot_table(index="ID", columns="Category", values="Values")
ddf.columns = list(ddf.columns)
ddf.to_parquet("./")

#shutdown commented for further testing
#client.shutdown()
```

</p>
</details>

Already high memory usage during drop_duplicates:
<img width="950" alt="dask_drop_duplicates" src="https://user-images.githubusercontent.com/37546799/92284212-e25d6a00-ef01-11ea-8542-db689119a946.png">

Memory usage >400% during pivot_table:
<img width="954" alt="dask_pivot_table" src="https://user-images.githubusercontent.com/37546799/92284210-e25d6a00-ef01-11ea-98f4-06d5650cc582.png">

Jupyter notebook log indicating memory leak:
<img width="668" alt="dask_memleak" src="https://user-images.githubusercontent.com/37546799/92284209-e12c3d00-ef01-11ea-9abf-01d2639658ab.png">

Code in form of Jupyter notebook for convenience
[24 Minimum Reproducible Example Ext.zip](https://github.com/dask/dask/files/5176969/24.Minimum.Reproducible.Example.Ext.zip)

**Comment by klnrdknt** [comment]
I am also looking for a solution on this. Exactly what is the recipe using groupby instead of pivot_table? Is this equivalent?

**Comment by martindurant** [comment]
Given the lack of activity here, I will close this in the absence of a clear reproducer and the passage of time. Please raise a *new* issue with a reproducer if the problem arises again.

**Comment by LutzFassl** [comment]
What

> Given the lack of activity here, I will close this in the absence of a clear reproducer and the passage of time. Please raise a _new_ issue with a reproducer if the problem arises again.

What was the problem with the code example that I provided? It’s not exactly minimal, but should be complete and reproduce the error (at least on systems with ram size <=24gig). The problem shows up only at a certain size of data therefore I needed/wanted to create a sizeable amount first. But could also just download a file instead of generating it.

**Comment by martindurant** [comment]
My apologies, your code was right there, but in an expandable block.

With the current version of dask, I am finding that the process *almost* completes, with "data stored" per worker hovering around 4GB. There are no longer warnings about memory leaks, at least for me. Perhaps a little more memory per worker would have been enough.

You should be aware that pivot_table produces a *single* partition of output. With the many stages for processing (load, as-type, categorize, reindex, shuffle, aggregate), there are MANY intermediate copies of the data in flight at once, all of which will need to be materialised into that final partition (and pandas `concat` is known not to be particularly memory efficient...). I also note that the small choice of partition chunk size means that there are 1000 tasks in this graph.

Im afraid, 'm not sure what the output should look like, so I don't know how you would have done this with groupby.

**Comment by LutzFassl** [comment]
Thanks for your reply! I now understand that it's a given that pivot_table materialises into one partition. I was under the impression that dask should also carry this out in a distributed fashion. I worked around the problem eventually. Do you know if there is any plan to carry out pivot_table distributed in the future?

To your last comment: In this code example it's really the most classical / "standard" pivoting task: A generic (index, column, value).  **_To substitute the pivot_table with groupby would most probably solve my issue!_** - Thanks for this idea!

If I run it again, I also noted from your answer to try:
- current version, now that half a year hassed there seem to be a bunch of updates
- bigger block size and or increased memory_limit
